### PR TITLE
OKTA-140057 : Initial

### PR DESCRIPTION
**Pull request by ericcarroll-okta**
This is the initial check-in of the ADFS Adapter, plus unit tests.

Primary Reviewer: @nandagopalseshagiri-okta 
Optional Reviewer: @dannavarro-okta 

## Screenshot of  ADFS Sign In hosting Okta Sign In Widget ##
screen shot 2017-11-02 at 12 17 42 pm

## Screenshot of ADFS displaying an error from the adapter ##
screen shot 2017-11-02 at 2 45 10 pm

## Screenshot of how adapter is selected in the AD FS Management console ##
adfs_edit_policy

### CODE REVIEW

File: OktaMfaAdfs/AuthenticationAdapter.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            // AuthN either doesn't know this user, or this user doesn't support MFA
+            return false;
+        }
+
+        /// <summary>
+        /// Called by ADFS when it determines MFA is required and available for a user.
+        /// </summary>
+        /// <param name="identityClaim">We get the username from here</param>
+        /// <param name="request">The request</param>
+        /// <param name="context">In the call to IsAvailableForUser, we stored the JWT post response here</param>
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/AuthenticationAdapter.cs]: is `IsAvailableForUser` called by ADFS?

File: OktaMfaAdfs/AuthenticationAdapter.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        public IAdapterPresentation TryEndAuthentication(IAuthenticationContext context, IProofData proofData, HttpListenerRequest request, out Claim[] claims)
+        {
+            try
+            {
+                // confirm the user completed MFA successfully
+                string stateToken = context.Data[STATE_TOKEN_KEY] as string;
+                AuthNService.Response stateTokenResponse = AuthNService.PostStateToken(RequestFactory, config.Url, stateToken);
+                if (stateTokenResponse.IsSuccess)
+                {
+                    claims = new Claim[] { new Claim(ClaimTypes.AuthenticationMethod, AuthenticationAdapterMetadata.AUTH_METHOD_OTP) };
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/AuthenticationAdapter.cs]: Is it necessarily OTP?

File: OktaMfaAdfs/AuthenticationAdapterMetadata.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        /// </summary>
+        public Dictionary<int, string> Descriptions => new Dictionary<int, string>
+                {
+                    { 1033, ADMIN_NAME }
+                };
+
+        /// <summary>
+        /// List of supported authentication methods
+        /// </summary>
+        public string[] AuthenticationMethods => new String[] { AUTH_METHOD_OTP };
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/AuthenticationAdapterMetadata.cs]: why is it just otp? couldn't it also be a push or a hard token?

File: OktaMfaAdfs/Resources/SignInWidgetContainer.html [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+<form method='post' id='adapterGlue' autocomplete='off'>
+	<input id='context' type='hidden' name='Context' value='%Context%'/>
+	<input id='authMethod' type='hidden' name='AuthMethod' value='%AuthMethod%'/>
+</form>
+
+<script>
+	var orgUrl = '%BASE_URL%';
+	var stateToken = '%STATE_TOKEN%';
+    var oktaSignIn = new OktaSignIn({
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/Resources/SignInWidgetContainer.html]: nit: indentation

File: OktaMfaAdfs/Resources/SignInWidgetContainer.html [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+		features: {
+			hideSignOutLinkInMFA: true
+		}
+	});
+
+    oktaSignIn.renderEl(
+      { el: '#okta-login-container' },
+      function (res) {
+        if (res.status === 'SUCCESS') {
+			document.getElementById('adapterGlue').submit();
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/Resources/SignInWidgetContainer.html]: nit: indentation

File: OktaMfaAdfs/Resources/SignInWidgetContainer.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+		}
+      }
+    );
+</script>
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/Resources/SignInWidgetContainer.html]: nit: should include endline

File: OktaMfaAdfs/config.json [hunk 7]
```diff
@@ -0,0 +1 @@
+{ "url":"https://acme.okta1.com", "clientId":"0oa34lqe7ZbvtmvTH0g4", "clientSecret": "3VWzA5gkQIX8UDXNRGr5AATRit_4s0epA-2be09I" }
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/config.json]: shouldn't this not be included?

File: OktaMfaAdfs/AdapterPresentation.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using Microsoft.IdentityServer.Web.Authentication.External;
+
+namespace OktaMfaAdfs
+{
+    public abstract class AdapterPresentation : IAdapterPresentation, IAdapterPresentationForm
+    {
+        const string PAGE_TITLE = "Okta MFA";
```
[the remark this answers is not in the document]
Comment (ericcarroll-okta) [reply OktaMfaAdfs/AdapterPresentation.cs]: I've got a list of stuff like this for James.

File: OktaMfaAdfs/AdapterPresentationWithPermanentFailure.cs [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+        private string errorMessage;
+
+        /// <summary>
+        /// Return HTML to imbed a permanent failure in the authentication page.
+        /// </summary>
+        /// <param name="lcid">Language</param>
+        /// <returns>HTML to imbed failure</returns>
+        protected override string GetFormHtmlImpl(int lcid)
+        {
+            string html = string.Format("<p>{0}</p>", errorMessage);
```
[the remark this answers is not in the document]
Comment (ericcarroll-okta) [reply OktaMfaAdfs/AdapterPresentationWithPermanentFailure.cs]: This is another item on the James list.  But I'll add a screenshot of what it looks like now.

File: OktaMfaAdfs/AuthNService.cs [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            return new Response(JsonConvert.DeserializeObject<Dictionary<string, dynamic>>(webResponse.Body));
+        }
+
+        public class Response
+        {
+            private const string STATE_TOKEN_KEY = "stateToken";
+            private const string STATUS_KEY = "status";
+
+            private const string SUCCESS_STATUS = "SUCCESS";
+            private const string MFA_REQUIRED_STATUS = "MFA_REQUIRED";
```
[the remark this answers is not in the document]
Comment (ericcarroll-okta) [reply OktaMfaAdfs/AuthNService.cs]: Wasn't aware of it.  I'll look into it.

File: OktaMfaAdfs/AuthenticationAdapter.cs (hunk 1 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/AuthenticationAdapter.cs]: Yes.  When ADFS determines a user needs MFA, it calls this function to see if the adapter can do MFA for this user.  If the user is unknown to Okta, or not set to use MFA in Okta, the adapter returns false.  Which looks like an error condition to the user.

I talked with Alex and James about how this should work, given both ADFS and Okta have sign-in policies.  The agreement is admins should use ADFS policies, and Okta's MFA Only App should say yes to MFA for anyone ADFS asks about.

File: OktaMfaAdfs/AuthenticationAdapter.cs (hunk 2 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/AuthenticationAdapter.cs]: I have this on my list to investigate.  I haven't figured out how ADFS uses it yet.

File: OktaMfaAdfs/AuthenticationAdapterMetadata.cs (hunk 3 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/AuthenticationAdapterMetadata.cs]: Also being investigated.

File: OktaMfaAdfs/Resources/SignInWidgetContainer.html (hunk 4 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/Resources/SignInWidgetContainer.html]: I'll see what I can do - looks fine in Visual Studio 😄

File: OktaMfaAdfs/config.json (hunk 7 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/config.json]: It won't be used by the installer, but it's useful for setting up on a dev box.

File: OktaMfaAdfs/AuthenticationAdapter.cs (hunk 1 above)
Comment (dannavarro-okta) [reply OktaMfaAdfs/AuthenticationAdapter.cs]: do you know what happens if we send a fake user to authn on the JWT? Will the state token return deny or something?

File: OktaMfaAdfs/AuthenticationAdapter.cs (hunk 2 above)
Comment (dannavarro-okta) [reply OktaMfaAdfs/AuthenticationAdapter.cs]: I don't think that claim is relevant for us. Based on [this blog](https://dirteam.com/sander/2017/01/16/forcing-the-use-of-a-specific-azure-multi-factor-authentication-method-for-a-relying-party-trust-in-ad-fs/) I think it is only used for forcing a certain type of Azure MFA. But in this case Okta is handling the MFA, not Azure

**Comment by dannavarro-okta** [comment]
seems like you might be checking in some extra files. Are they all needed?

File: OktaMfaAdfs/AuthenticationAdapter.cs (hunk 1 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/AuthenticationAdapter.cs]: The adapter is only invoked after successfully signing into ADFS.  If you sign in as a user that is unknown to Okta, the _authn_ call returns a not-found/404.  The adapter then returns _false_ from _IsAvailableForUser_.

**Comment by ericcarroll-okta** [comment]
Any file in particular 😄?  I google for what I should check in, and mostly followed that.  There are some file for ease of setting up the dev machine:  config.json and the scripts.  There's the reference DLL for the adapter - don't think VS can just download that.

File: OktaMfaAdfs/AuthenticationAdapter.cs [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+                    context.Data.Add(STATE_TOKEN_KEY, jwtResponse.StateToken);
+                    return true;
+                }
+            }
+            catch (Exception)
+            {
+                // fall through to below.  ADFS logs this exception.
+            }
+
+            // AuthN either doesn't know this user, or this user doesn't support MFA
```
Comment (nandagopalseshagiri-okta) [inline OktaMfaAdfs/AuthenticationAdapter.cs]: We would have to log to a file inside the log folder inside the location where the ADFS plugin resides. We can tackle that in a separate JIRA.

File: OktaMfaAdfs/AuthNService.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (dannavarro-okta) [reply OktaMfaAdfs/AuthNService.cs]: @nandagopalseshagiri-okta should we be handling all `MFA_*` statuses?

File: OktaMfaAdfs/AuthenticationAdapter.cs (hunk 11 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/AuthenticationAdapter.cs]: Returning _false_ from the above puts an error into the ADFS event log.  There are tools to gather these event logs into one spot, so this is probably where we want to put stuff.

I took a qucik look at adding events directly to the log, so we could provide more details.  Didn't find anything, but I'll take another look.

Comment (ericcarroll-okta) [reply OktaMfaAdfs/AuthenticationAdapter.cs]: Looks like I can register the adapter to write events to the server's event log, under a unique name.  Meaning, I cannot log events for ADFS, but I can log events for ADFS Okta Adapter (or whatever).

**Comment by ericcarroll-okta** [comment]
- James gave me updated text.
- Made a pass at improving the tab usage in the HTML file - better.
- Added a pair of missing unit tests for presentation classes.

File: OktaMfaAdfs.sln [hunk 12]
```diff
@@ -0,0 +1,31 @@
+﻿
```
Comment (dannavarro-okta) [inline OktaMfaAdfs.sln]: Is the sln file needed?

File: OktaMfaAdfs/JwtService.cs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        /// <summary>
+        /// Makes a JWT token
+        /// </summary>
+        /// <param name="timeProvider">Source of unix time values</param>
+        /// <param name="audience">Uri for the token audience, ex: https://acme.okta1.com/api/v1/authn</param>
+        /// <param name="clientId">The token issuer, comes from the Microsoft ADFS app settings</param>
+        /// <param name="clientSecret">Used to sign the token, comes from the Microsoft ADFS app settings</param>
+        /// <param name="username">The token subject, comes from the ADFS claim</param>
+        /// <returns>Token</returns>
+        public static string MintToken(IUnixTimeProvider timeProvider, string audience, string clientId, string clientSecret, string username)
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/JwtService.cs]: seems a little weird to me to pass the timeProvider instead of the time

File: OktaMfaAdfs/OktaMfaAdfs.csproj [hunk 14]
```diff
@@ -0,0 +1,101 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/OktaMfaAdfs.csproj]: is this file needed

File: OktaMfaAdfs/Properties/AssemblyInfo.cs [hunk 15]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/Properties/AssemblyInfo.cs]: is this file neeeded?

File: OktaMfaAdfs/Properties/Resources.Designer.cs [hunk 16]
```diff
@@ -0,0 +1,77 @@
+﻿//------------------------------------------------------------------------------
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/Properties/Resources.Designer.cs]: needed?

File: OktaMfaAdfs/Properties/Resources.resx [hunk 17]
```diff
@@ -0,0 +1,124 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/Properties/Resources.resx]: needed?

File: OktaMfaAdfs/Resources/SignInWidgetContainer.html [hunk 18]
```diff
@@ -0,0 +1,31 @@
+﻿<script src="https://ok1static.oktacdn.com/assets/js/sdk/okta-signin-widget/2.1.0/js/okta-sign-in.min.js" type="text/javascript"></script>
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/Resources/SignInWidgetContainer.html]: maybe want to get Mau to sign off on this

File: OktaMfaAdfs/UnixTimeProviderImpl.cs [hunk 19]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/UnixTimeProviderImpl.cs]: this class is pretty simple, why did you split it between an interface and an impl?

File: OktaMfaAdfs/app.config [hunk 20]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/app.config]: needed?

File: OktaMfaAdfs/packages.config [hunk 21]
```diff
@@ -0,0 +1,7 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/packages.config]: needed?

File: OktaMfaAdfs/packages.config [hunk 22]
```diff
@@ -0,0 +1,7 @@
...
+<packages>
+  <package id="Microsoft.IdentityModel.Logging" version="1.1.4" targetFramework="net461" />
+  <package id="Microsoft.IdentityModel.Tokens" version="5.1.4" targetFramework="net461" />
+  <package id="Newtonsoft.Json" version="9.0.1" targetFramework="net461" />
+  <package id="System.IdentityModel.Tokens.Jwt" version="5.1.4" targetFramework="net461" />
+</packages>
```
Comment (dannavarro-okta) [inline OktaMfaAdfs/packages.config]: should include endline

File: OktaMfaAdfsTest/Helpers/AuthenticationMethodConfigDataForTests.cs [hunk 23]
```diff
@@ -0,0 +1,19 @@
+﻿using System.IO;
```
Comment (dannavarro-okta) [inline OktaMfaAdfsTest/Helpers/AuthenticationMethodConfigDataForTests.cs]: Do these *ForTests classes need to be named differently from the main classes? I'm assuming you can't use the main classes because they are in a different directory so why not keep the naming consistent or find a common place for them where they can be shared?

File: OktaMfaAdfsTest/Properties/AssemblyInfo.cs [hunk 24]
```diff
@@ -0,0 +1,20 @@
+using System.Reflection;
```
Comment (dannavarro-okta) [inline OktaMfaAdfsTest/Properties/AssemblyInfo.cs]: needed?

File: OktaMfaAdfsTest/packages.config [hunk 25]
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (dannavarro-okta) [inline OktaMfaAdfsTest/packages.config]: needed?

File: OktaMfaAdfs.sln (hunk 12 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs.sln]: There isn't much documentation on this, but what I did find said to put it in source control.  It doesn't have any local-install stuff (ex: the paths in the file are relative).  Double checked the winlogin-agent repo, and they checked them in too.

File: OktaMfaAdfs/OktaMfaAdfs.csproj (hunk 14 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/OktaMfaAdfs.csproj]: Same answer as the SLN - not part of the suggested gitignore, and included by winlogin-agent.

File: OktaMfaAdfs/Properties/Resources.Designer.cs (hunk 16 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/Properties/Resources.Designer.cs]: Not in the suggested gitignore.  Also not in winlogin-agent.  It does get regenerated as needed, but folks suggest adding it to source control.

File: OktaMfaAdfs/Properties/Resources.resx (hunk 17 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/Properties/Resources.resx]: Same answer again 😄

File: OktaMfaAdfs/Resources/SignInWidgetContainer.html (hunk 18 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/Resources/SignInWidgetContainer.html]: Visual Studio has odd ideas... I'll see what I can do.

I'll ask Mau if he wants to review.

File: OktaMfaAdfs/UnixTimeProviderImpl.cs (hunk 19 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/UnixTimeProviderImpl.cs]: So I could inject a test version for unit tests - one that gives me a predicable answer.  Primitive mocking, since VS doesn't provide it.

File: OktaMfaAdfs/app.config (hunk 20 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/app.config]: Same answer again 😄

File: OktaMfaAdfs/packages.config (hunk 21 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/packages.config]: Same answer again 😄

File: OktaMfaAdfs/packages.config (hunk 22 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfs/packages.config]: This is an auto-generated file

File: OktaMfaAdfsTest/Helpers/AuthenticationMethodConfigDataForTests.cs (hunk 23 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfsTest/Helpers/AuthenticationMethodConfigDataForTests.cs]: I wanted the name to make it clear these are mock versions.  If I gave them the same name they could conflict - the tests have _using_ statements for the not-time implementations.

I guess I could replace _ForTests_ with _Mock_

File: OktaMfaAdfsTest/Properties/AssemblyInfo.cs (hunk 24 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfsTest/Properties/AssemblyInfo.cs]: Same answer again 😄

File: OktaMfaAdfsTest/packages.config (hunk 25 above)
Comment (ericcarroll-okta) [reply OktaMfaAdfsTest/packages.config]: Yup!

**Comment by ericcarroll-okta** [comment]
There are two binary files included in the project:
* OktaMfaAdfsKey.pfx : Needed to put a DLL in the GAC. Used to build the provider.
* Microsfot.IdentityServer.Web.dll : Reference DLL for the provider interface.  I'll double-check it cannot be referenced the normal way.

I checked all the file types for inclusion in source control.  Found GitHub's suggested gitignore which helped bunches.  Wish I'd found it earlier 😄 .
